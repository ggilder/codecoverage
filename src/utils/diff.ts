interface FileDiff {
  filename: string
  addedLines: number[]
  deletedLines: number[]
}

export function parseGitDiff(diffOutput: string): FileDiff[] {
  const fileDiffs: FileDiff[] = []
  const lines = diffOutput.split('\n')

  let currentFileDiff: FileDiff | undefined
  let currentAddedLines: number[] = []
  let currentDeletedLines: number[] = []
  let seenHeaderLine = false
  let deletionCurrentLineNumber = 0
  let additionCurrentLineNumber = 0

  for (const line of lines) {
    if (line.startsWith('diff --git')) {
      // New file diff starts
      if (currentFileDiff) {
        currentFileDiff.addedLines = currentAddedLines
        currentFileDiff.deletedLines = currentDeletedLines
        fileDiffs.push(currentFileDiff)
      }

      currentFileDiff = {
        filename: getFilenameFromDiffHeader(line),
        addedLines: [],
        deletedLines: []
      }
      currentAddedLines = []
      currentDeletedLines = []
      seenHeaderLine = false
    } else if (line.startsWith('@@')) {
      // Header line
      seenHeaderLine = true
      const lineInfo = getLineInfoFromHeaderLine(line)
      deletionCurrentLineNumber = lineInfo.deletionStartingLineNumber
      additionCurrentLineNumber = lineInfo.additionStartingLineNumber
    } else if (line.startsWith('+') && seenHeaderLine) {
      // Added line
      currentAddedLines.push(additionCurrentLineNumber)
      additionCurrentLineNumber++
    } else if (line.startsWith('-') && seenHeaderLine) {
      // Deleted line
      currentDeletedLines.push(deletionCurrentLineNumber)
      deletionCurrentLineNumber++
    } else if (seenHeaderLine) {
      // Context line
      deletionCurrentLineNumber++
      additionCurrentLineNumber++
    }
  }

  // Add the last file diff
  if (currentFileDiff) {
    currentFileDiff.addedLines = currentAddedLines
    currentFileDiff.deletedLines = currentDeletedLines
    fileDiffs.push(currentFileDiff)
  }

  return fileDiffs
}

function getFilenameFromDiffHeader(header: string): string {
  // Extract the filename from the diff header
  const startIndex = header.indexOf(' a/') + 3
  const endIndex = header.indexOf(' b/', startIndex)
  const filename = header.substring(startIndex, endIndex)
  return filename
}

function getLineInfoFromHeaderLine(line: string): {
  deletionStartingLineNumber: number
  additionStartingLineNumber: number
} {
  // Extract the starting line numbers for each side of the diff
  const matches = line.match(/-(\d+),?(\d+)? \+(\d+),?(\d+)? @@/)
  if (matches && matches.length === 5) {
    const deletionStartingLineNumber = parseInt(matches[1], 10)
    const additionStartingLineNumber = parseInt(matches[3], 10)
    return {deletionStartingLineNumber, additionStartingLineNumber}
  }
  return {deletionStartingLineNumber: 0, additionStartingLineNumber: 0}
}
