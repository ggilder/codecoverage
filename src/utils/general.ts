export function filterCoverageByFile(coverage: CoverageParsed): CoverageFile[] {
  return coverage.map(item => ({
    fileName: item.file,
    missingLineNumbers: item?.lines?.details
      .filter(line => line.hit === 0)
      .map(line => line.line)
  }))
}

export function coalesceLineNumbers(lines: number[]): LineRange[] {
  const ranges: LineRange[] = []
  let rstart
  let rend
  for (let i = 0; i < lines.length; i++) {
    rstart = lines[i]
    rend = rstart
    while (lines[i + 1] - lines[i] === 1) {
      rend = lines[i + 1]
      i++
    }
    ranges.push({start_line: rstart, end_line: rend})
  }
  return ranges
}

export function intersectLineRanges(
  rangesA,
  rangesB: LineRange[]
): LineRange[] {
  const outRanges: LineRange[] = []
  for (const bRange of rangesB) {
    const aRangeIntersects = rangesA.filter(aRange => {
      return (
        (bRange.start_line >= aRange.start_line &&
          bRange.start_line <= aRange.end_line) ||
        (bRange.end_line >= aRange.start_line &&
          bRange.end_line <= aRange.end_line) ||
        (aRange.start_line >= bRange.start_line &&
          aRange.start_line <= bRange.end_line) ||
        (aRange.end_line >= bRange.start_line &&
          aRange.end_line <= bRange.end_line)
      )
    })
    for (const aRange of aRangeIntersects) {
      outRanges.push({
        start_line: Math.max(aRange.start_line, bRange.start_line),
        end_line: Math.min(aRange.end_line, bRange.end_line)
      })
    }
  }
  return outRanges
}

export type CoverageParsed = {
  file: string
  title: string
  lines: {
    found: number
    hit: number
    details: {
      line: number
      hit: number
      name: string
    }[]
  }
}[]

export type CoverageFile = {
  fileName: string
  missingLineNumbers: number[]
}

export type LineRange = {
  start_line: number
  end_line: number
}
