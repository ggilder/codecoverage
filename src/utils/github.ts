import * as core from '@actions/core'
import * as diff from './diff'
import * as github from '@actions/github'
import {
  CoverageFile,
  LineRange,
  coalesceLineNumbers,
  intersectLineRanges
} from './general'
import {Octokit} from 'octokit'

export class GithubUtil {
  private client: Octokit

  constructor(token: string) {
    if (!token) {
      throw new Error('GITHUB_TOKEN is missing')
    }
    this.client = new Octokit({auth: token})
  }

  getPullRequestRef(): string {
    const pullRequest = github.context.payload.pull_request
    return pullRequest
      ? pullRequest.head.ref
      : github.context.ref.replace('refs/heads/', '')
  }

  async getPullRequestDiff(): Promise<PullRequestFiles> {
    const pull_number = github.context.issue.number
    const response = await this.client.rest.pulls.get({
      ...github.context.repo,
      pull_number,
      mediaType: {
        format: 'diff'
      }
    })
    // @ts-expect-error With mediaType param, response.data is actually a string, but the response type doesn't reflect this
    const fileLines = diff.parseGitDiff(response.data)
    const prFiles: PullRequestFiles = {}
    for (const item of fileLines) {
      prFiles[item.filename] = coalesceLineNumbers(item.addedLines)
    }

    return prFiles
  }

  /**
   * https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-warning-message
   */
  annotate(annotations: Annotations[]): void {
    for (const ann of annotations) {
      console.log(
        `::${ann.annotation_level} file=${ann.path},line=${ann.start_line},endLine=${ann.end_line}::${ann.message}`
      )
    }
  }

  buildAnnotations(
    coverageFiles: CoverageFile[],
    pullRequestFiles: PullRequestFiles
  ): Annotations[] {
    const annotations: Annotations[] = []
    for (const current of coverageFiles) {
      // Only annotate relevant files
      const prFileRanges = pullRequestFiles[current.fileName]
      if (prFileRanges) {
        const coverageRanges = coalesceLineNumbers(current.missingLineNumbers)
        const uncoveredRanges = intersectLineRanges(
          coverageRanges,
          prFileRanges
        )

        // Only annotate relevant line ranges
        for (const uRange of uncoveredRanges) {
          const message =
            uRange.end_line > uRange.start_line
              ? 'These lines are not covered by a test'
              : 'This line is not covered by a test'
          annotations.push({
            path: current.fileName,
            start_line: uRange.start_line,
            end_line: uRange.end_line,
            annotation_level: 'warning',
            message
          })
        }
      }
    }
    core.info(`Annotation count: ${annotations.length}`)
    return annotations
  }
}

type Annotations = {
  path: string
  start_line: number
  end_line: number
  start_column?: number
  end_column?: number
  annotation_level: string
  message: string
}

type PullRequestFiles = {
  [key: string]: LineRange[]
}
