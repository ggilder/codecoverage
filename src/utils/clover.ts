import * as clover from '@cvrg-report/clover-json'
import * as fs from 'fs'
import * as path from 'path'
import {CoverageParsed} from './general'

export async function parseClover(
  coveragePath: string,
  workspacePath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Clover XML path provided')
  }
  if (!workspacePath) {
    throw Error('No workspace path provided')
  }
  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  const parsed = await clover.parseContent(fileRaw)

  for (const entry of parsed) {
    entry.file = path.relative(workspacePath, entry.path || entry.file)
  }

  return parsed
}
