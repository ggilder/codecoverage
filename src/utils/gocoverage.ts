import * as fs from 'fs'
import * as gocov from 'golang-cover-parse'
import * as path from 'path'
import * as readline from 'readline'
import {CoverageParsed} from './general'

export async function parseGoCoverage(
  coveragePath: string,
  goModPath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Go coverage path provided')
  }

  if (!goModPath) {
    throw Error('No Go module path provided')
  }

  const goModule = await parseGoModFile(goModPath)

  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  return new Promise((resolve, reject) => {
    gocov.parseContent(fileRaw, (err, result) => {
      if (err === null) {
        filterModulePaths(result, goModule)
        resolve(result)
      }
      reject(err)
    })
  })
}

function filterModulePaths(entries, moduleName: string): void {
  for (const entry of entries) {
    entry.file = path.relative(moduleName, entry.file)
  }
}

async function parseGoModFile(filePath: string): Promise<string> {
  const fileStream = fs.createReadStream(filePath)
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  })
  // Note: we use the crlfDelay option to recognize all instances of CR LF
  // ('\r\n') in input.txt as a single line break.

  for await (const line of rl) {
    if (line.startsWith('module ')) {
      return line.slice(7)
    }
  }

  /* istanbul ignore next */
  return ''
}
