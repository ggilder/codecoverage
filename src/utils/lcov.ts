import * as NodeUtil from 'util'
import * as fs from 'fs'
import * as path from 'path'
import {CoverageParsed} from './general'
import parser from 'lcov-parse'

export async function parseLCov(
  lcovPath: string,
  workspacePath: string
): Promise<CoverageParsed> {
  if (!lcovPath) {
    throw Error('No LCov path provided')
  }

  const parserPromise = NodeUtil.promisify(parser)
  const fileRaw = fs.readFileSync(lcovPath, 'utf8')
  const parsed = (await parserPromise(fileRaw)) as CoverageParsed

  for (const entry of parsed) {
    entry.file = path.relative(workspacePath, entry.file)
  }

  return parsed
}
