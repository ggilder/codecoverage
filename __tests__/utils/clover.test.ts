import {test} from '@jest/globals'
import {parseClover} from '../../src/utils/clover'
import {getFixturePath} from '../fixtures/util'

test('should parse Clover file', async function () {
  const path = getFixturePath('clover.xml')
  const output = await parseClover(path, '/var/www/html')
  expect(output).toMatchSnapshot()
})

test('should parse Clover file with paths', async function () {
  const path = getFixturePath('clover_with_path.xml')
  const output = await parseClover(path, '/Users/gabriel/src/project')
  expect(output).toMatchSnapshot()
})

test('should throw err if file path is not given', async function () {
  await expect(parseClover('', '')).rejects.toThrow(
    'No Clover XML path provided'
  )
})

test('should throw err if workspace path is not given', async function () {
  await expect(parseClover('clover.xml', '')).rejects.toThrow(
    'No workspace path provided'
  )
})
