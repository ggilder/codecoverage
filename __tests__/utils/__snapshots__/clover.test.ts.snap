// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse Clover file 1`] = `
Array [
  Object {
    "file": "app/Actions/MyAction.php",
    "functions": Object {
      "details": Array [
        Object {
          "hit": 15,
          "line": 21,
          "name": "__construct",
        },
        Object {
          "hit": 14,
          "line": 33,
          "name": "execute",
        },
      ],
      "found": 2,
      "hit": 2,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 15,
          "line": 23,
        },
        Object {
          "hit": 15,
          "line": 24,
        },
        Object {
          "hit": 15,
          "line": 25,
        },
        Object {
          "hit": 15,
          "line": 26,
        },
        Object {
          "hit": 15,
          "line": 27,
        },
        Object {
          "hit": 14,
          "line": 36,
        },
        Object {
          "hit": 14,
          "line": 37,
        },
        Object {
          "hit": 1,
          "line": 38,
        },
        Object {
          "hit": 1,
          "line": 39,
        },
        Object {
          "hit": 14,
          "line": 42,
        },
        Object {
          "hit": 14,
          "line": 43,
        },
        Object {
          "hit": 14,
          "line": 44,
        },
        Object {
          "hit": 14,
          "line": 45,
        },
        Object {
          "hit": 14,
          "line": 46,
        },
        Object {
          "hit": 14,
          "line": 47,
        },
        Object {
          "hit": 14,
          "line": 50,
        },
        Object {
          "hit": 14,
          "line": 51,
        },
        Object {
          "hit": 1,
          "line": 52,
        },
        Object {
          "hit": 1,
          "line": 53,
        },
        Object {
          "hit": 1,
          "line": 54,
        },
        Object {
          "hit": 1,
          "line": 55,
        },
        Object {
          "hit": 1,
          "line": 56,
        },
        Object {
          "hit": 1,
          "line": 57,
        },
        Object {
          "hit": 13,
          "line": 61,
        },
        Object {
          "hit": 13,
          "line": 65,
        },
        Object {
          "hit": 13,
          "line": 67,
        },
        Object {
          "hit": 13,
          "line": 68,
        },
        Object {
          "hit": 13,
          "line": 69,
        },
        Object {
          "hit": 13,
          "line": 70,
        },
        Object {
          "hit": 13,
          "line": 71,
        },
        Object {
          "hit": 13,
          "line": 72,
        },
        Object {
          "hit": 13,
          "line": 74,
        },
        Object {
          "hit": 13,
          "line": 75,
        },
        Object {
          "hit": 0,
          "line": 78,
        },
      ],
      "found": 34,
      "hit": 33,
    },
    "path": undefined,
    "title": "App\\\\Actions\\\\MyAction",
  },
  Object {
    "file": "app/Console/Kernel.php",
    "functions": Object {
      "details": Array [
        Object {
          "hit": 13,
          "line": 26,
          "name": "schedule",
        },
      ],
      "found": 1,
      "hit": 1,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 13,
          "line": 29,
        },
      ],
      "found": 1,
      "hit": 1,
    },
    "path": undefined,
    "title": "App\\\\Console\\\\Kernel",
  },
  Object {
    "file": "app/Events/MyEvent.php",
    "functions": Object {
      "details": Array [
        Object {
          "hit": 0,
          "line": 16,
          "name": "__construct",
        },
      ],
      "found": 1,
      "hit": 0,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 0,
          "line": 18,
        },
        Object {
          "hit": 0,
          "line": 19,
        },
      ],
      "found": 2,
      "hit": 0,
    },
    "path": undefined,
    "title": "App\\\\Events\\\\MyEvent",
  },
]
`;

exports[`should parse Clover file with paths 1`] = `
Array [
  Object {
    "file": "sub/app.config.ts",
    "functions": Object {
      "details": Array [],
      "found": 0,
      "hit": 0,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 1,
          "line": 1,
        },
        Object {
          "hit": 1,
          "line": 2,
        },
        Object {
          "hit": 1,
          "line": 3,
        },
        Object {
          "hit": 1,
          "line": 4,
        },
        Object {
          "hit": 1,
          "line": 5,
        },
        Object {
          "hit": 1,
          "line": 6,
        },
        Object {
          "hit": 1,
          "line": 7,
        },
        Object {
          "hit": 1,
          "line": 8,
        },
        Object {
          "hit": 1,
          "line": 9,
        },
      ],
      "found": 9,
      "hit": 9,
    },
    "path": "/Users/gabriel/src/project/sub/app.config.ts",
    "title": null,
  },
  Object {
    "file": "sub/app.vue",
    "functions": Object {
      "details": Array [],
      "found": 0,
      "hit": 0,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 1,
          "line": 1,
        },
        Object {
          "hit": 1,
          "line": 2,
        },
        Object {
          "hit": 1,
          "line": 3,
        },
        Object {
          "hit": 1,
          "line": 4,
        },
        Object {
          "hit": 1,
          "line": 5,
        },
        Object {
          "hit": 1,
          "line": 6,
        },
        Object {
          "hit": 1,
          "line": 7,
        },
        Object {
          "hit": 1,
          "line": 8,
        },
        Object {
          "hit": 1,
          "line": 9,
        },
        Object {
          "hit": 1,
          "line": 10,
        },
        Object {
          "hit": 1,
          "line": 11,
        },
        Object {
          "hit": 1,
          "line": 12,
        },
        Object {
          "hit": 1,
          "line": 13,
        },
        Object {
          "hit": 1,
          "line": 14,
        },
        Object {
          "hit": 1,
          "line": 15,
        },
        Object {
          "hit": 1,
          "line": 16,
        },
        Object {
          "hit": 1,
          "line": 17,
        },
        Object {
          "hit": 1,
          "line": 18,
        },
        Object {
          "hit": 1,
          "line": 19,
        },
        Object {
          "hit": 1,
          "line": 20,
        },
        Object {
          "hit": 1,
          "line": 21,
        },
        Object {
          "hit": 1,
          "line": 22,
        },
        Object {
          "hit": 1,
          "line": 23,
        },
        Object {
          "hit": 1,
          "line": 24,
        },
        Object {
          "hit": 1,
          "line": 25,
        },
        Object {
          "hit": 1,
          "line": 26,
        },
        Object {
          "hit": 1,
          "line": 27,
        },
        Object {
          "hit": 1,
          "line": 28,
        },
        Object {
          "hit": 1,
          "line": 29,
        },
        Object {
          "hit": 1,
          "line": 30,
        },
        Object {
          "hit": 1,
          "line": 31,
        },
        Object {
          "hit": 1,
          "line": 32,
        },
        Object {
          "hit": 1,
          "line": 33,
        },
        Object {
          "hit": 1,
          "line": 34,
        },
        Object {
          "hit": 1,
          "line": 35,
        },
        Object {
          "hit": 1,
          "line": 36,
        },
        Object {
          "hit": 1,
          "line": 37,
        },
        Object {
          "hit": 1,
          "line": 38,
        },
        Object {
          "hit": 1,
          "line": 39,
        },
        Object {
          "hit": 1,
          "line": 40,
        },
        Object {
          "hit": 1,
          "line": 41,
        },
        Object {
          "hit": 1,
          "line": 42,
        },
        Object {
          "hit": 1,
          "line": 43,
        },
        Object {
          "hit": 1,
          "line": 44,
        },
        Object {
          "hit": 1,
          "line": 45,
        },
        Object {
          "hit": 1,
          "line": 46,
        },
        Object {
          "hit": 1,
          "line": 47,
        },
        Object {
          "hit": 1,
          "line": 48,
        },
        Object {
          "hit": 1,
          "line": 49,
        },
        Object {
          "hit": 1,
          "line": 50,
        },
        Object {
          "hit": 1,
          "line": 51,
        },
        Object {
          "hit": 1,
          "line": 52,
        },
        Object {
          "hit": 1,
          "line": 53,
        },
        Object {
          "hit": 1,
          "line": 54,
        },
        Object {
          "hit": 1,
          "line": 55,
        },
        Object {
          "hit": 1,
          "line": 56,
        },
        Object {
          "hit": 1,
          "line": 57,
        },
        Object {
          "hit": 1,
          "line": 58,
        },
        Object {
          "hit": 1,
          "line": 59,
        },
        Object {
          "hit": 1,
          "line": 60,
        },
        Object {
          "hit": 1,
          "line": 61,
        },
        Object {
          "hit": 1,
          "line": 62,
        },
        Object {
          "hit": 1,
          "line": 63,
        },
        Object {
          "hit": 1,
          "line": 64,
        },
        Object {
          "hit": 1,
          "line": 65,
        },
        Object {
          "hit": 1,
          "line": 66,
        },
        Object {
          "hit": 1,
          "line": 67,
        },
        Object {
          "hit": 1,
          "line": 68,
        },
        Object {
          "hit": 1,
          "line": 69,
        },
        Object {
          "hit": 1,
          "line": 70,
        },
        Object {
          "hit": 1,
          "line": 71,
        },
        Object {
          "hit": 1,
          "line": 72,
        },
        Object {
          "hit": 1,
          "line": 73,
        },
        Object {
          "hit": 1,
          "line": 74,
        },
        Object {
          "hit": 1,
          "line": 75,
        },
        Object {
          "hit": 1,
          "line": 76,
        },
        Object {
          "hit": 1,
          "line": 77,
        },
        Object {
          "hit": 1,
          "line": 78,
        },
        Object {
          "hit": 1,
          "line": 79,
        },
        Object {
          "hit": 1,
          "line": 80,
        },
        Object {
          "hit": 1,
          "line": 81,
        },
        Object {
          "hit": 1,
          "line": 82,
        },
        Object {
          "hit": 1,
          "line": 83,
        },
      ],
      "found": 83,
      "hit": 83,
    },
    "path": "/Users/gabriel/src/project/sub/app.vue",
    "title": null,
  },
]
`;
